src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs

```diff
@@ -137,6 +137,12 @@ public Uri Endpoint
         /// </summary>
         public string Headers { get; set; }
 
+        /// <summary>
+        /// Gets or sets optional headers factory for the connection. Refer to the <a href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#specifying-headers-via-environment-variables">
+        /// specification</a> for information on the expected format for Headers.
+        /// </summary>
+        public Func<string> HeadersFactory { get; set; }
```

Comment:
Thanks for this @DanikRaikhlin!

My thoughts...

We have factories other places and they are only called once. This factory seems to be called any time an export occurs? I think being inconsistent might be a problem. Also calling on every export might be overkill? It also isn't totally clear how `Headers` & `HeadersFactory` are supposed to work when specified together.

I was hoping to look at providing some generic mechanism in our exporters which could be used to reload them when options change (probably using `IOptionsMonitor` \ `IChangeToken`). I think that might be a more general-purpose fix for this (without some of the pitfalls) which could work across the board.

My vote would be let's hold off on this and look at some other options before choosing.

Reply:
Thanks for the feedback @CodeBlanch!

Fair point on factories being called once. Maybe the choice of naming it a factory wasn't the right one. The idea of keeping `Headers` and having a `HeadersFactory` is to avoid introducing a breaking change within existing api. I've considered `Headers` to be a static set of headers that are always added and `HeadersFactory` to be a resolver for additional headers that are calculated at runtime, thus resolved every call.

The usecase i'm facing is that the opentelemetry-collector has authentication enabled and expects an Authorization header which contains a bearer token that expires every N hours. As far as I can see there is no way I could hook into gRPC implementation to add an additional header at runtime. For http i did succeed by hooking into the IHttpClientFactory and configuring the HttpClient with a custom handler. But my goal is to get both gRPC and Http working. Or am I overseeing something that already makes this possible?

I like your idea of using IOptionMonitor and i've taken a quick look if it would be easy to implement. It feels like a big overall refactor would be needed to fully implement it everywhere. Also as a consumer I then would expect other options properties to be changable at runtime like HttpClientFactory which doesn't sound logical.

I have another idea I would like to pitch. What do you think about replacing `string Headers` with `IHeadersResolver HeadersResolver` within `OtlpExporterOptions`. Default value would have a default implementation of `IHeadersResolver` that contains current functionality but gives an extention point to consumer to change the returned values at runtime?

Reply:
@CodeBlanch thanks again for your time. I've updated the PR with the changes based on your feedback.

`OtlpExporterOptions` now contains a resolver function. Function parameter is the default headers collection that is resolved from Headers property. Te return type is also a headers collection that will actually be used during the requests. By default the function is configured to just return the parameter thus not breaking the shipped functionality. However it now gives the option to the consumer to expand upun the default behaviour or totally replace it. I've opted out my original suggestion of using a custom interface `IHeadersResolver` as it felt more of a service that could be used within dependency injection later on when DI is fully implemented. And the default behaviour of that service could still use the new function.

Reply:
> I would really like to NOT add this stuff to the options class. IMO "options" classes should just be POCOs which can be bound to config.

👍. Also, this options class is supposed to be reflective of the configuration options as defined by the specification: See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options.

We have a few additional things, namely export processor settings which were probably a mistake and we've discussed obsoleting. But that said, I would like to avoid having this class deviate further from the spec.

I agree, there's probably a better way to provide this kind of functionality via IConfiguration/IOptions.